### DISCUSSION TITLE
The plan for this repo

### DISCUSSION THREAD

**Discussion by frankelavsky** [fizzstudio/svg-aria-test#36] [General]
## The full testing plan for this repository
👋 Welcome! This is where we will record (and discuss) the SVG primitives, ARIA labels, screen readers, and browsers that we will be testing.

We’re using Discussions as a place separate from Issues, since this area is for moderators and maintainers only. This is a space where we can discuss our plans for this repo without using public issues. There also isn't a confusing use of "Issues" for things like project planning (which this is intended to be). Discussions about our testing matrices will go here, unless it is somehow related to an issue (then it goes in Issues).

This plan has 2 different (main) matrices, first is an intersection of SVG primitives we are testing and ARIA roles and states/properties. And then the second matrix is for every screen reader and browser combination we are doing. After these two matrices are tables that simply represent what each branch has for coverage specifically.

Obviously both of these main tables are not fully representative of every possible technology that we could test. However, the SVG/ARIA table is comprised of the roles and states/properties we would consider moderate or strong potential candidates for use with SVG. The Screen Reader/Browser table is a list of the most common browser/OS combos coupled with the most common screen readers used. Most of the data for this table comes from the [WebAim Screen Reader Survey from 2021](https://webaim.org/projects/screenreadersurvey9/#browsercombos).

### Limitations

Due to the massive complexity of possible testing scenarios across all SVG/ARIA and Screen Reader/Browser combinations, we have reduced our testing to the scope of the most common and most representative elements for our first pass of tests. All additional requests for testing should be considered with significant scrutiny as there is a strong risk for scope creep on this work if we are not careful.

### SVG / ARIA Matrix 

|                     | svg | g | circle | ellipsis | line | path | rect | text | title | desc |
| :----------------- | :-: |  :-:  | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| region              | ✔️ |   |   |   |   |   |   |   |   |   |
| group               |   | ✔️ |   |   |   |   |   |   |   |   | 
| img                 |   |   | ✔️ |   |   | ✔️ | ✔️ | ✔️ |   |   |
| graphics-document   | ✔️ |   |   |   |   |   |   |   |   |   |
| graphics-object     |   | ✔️ |   |   |   |   |   |   |   |   |
| graphics-symbol     |   |   | ✔️ |   |   | ✔️ | ✔️ | ✔️ |   |   |
| button              |   |   | ✔️ |   |   | ✔️ | ✔️ | ✔️ |   |   |
| menu                | ✔️ | ✔️ |   |   |   |   |   |   |   |   | 
| menuitem           |   | ✔️ |   |   |   |   |   |   |   |   |
| menuitemcheckbox  |   |   | ✔️ |   |   | ✔️ | ✔️ | ✔️ |   |   |
| aria-label       |   |   |   |   |   |   |   |   |   |   |
| aria-hidden       |   |   |   |   |   |   |   | ✔️ |   |   |
| aria-labelledby       | ✔️ | ✔️ | ✔️ |   |   | ✔️ | ✔️ | ✔️ |   |   |
| aria-describedby       |   |   |   |   |   |   |   |   |   |   |
| aria-expanded       |   | ✔️ |   |   |   |   |   |   |   |   |
| aria-haspopup       |   | ✔️ |   |   |   |   |   |   |   |   |
| aria-pressed        |   |   | ✔️ |   |   | ✔️ | ✔️ | ✔️ |   |   |
| aria-checked        |   |   | ✔️ |   |   | ✔️ | ✔️ | ✔️ |   |   |
| aria-live  |   |   |   |   |   |   |   | ✔️ |   |   |
| tabindex     |   |   | ✔️ |   |   | ✔️ | ✔️ | ✔️ |   |   |
| menubar |   |   |   |   |   |   |   |   |   |   |   
| application |   |   |   |   |   |   |   |   |   |   |   
| document |   |   |   |   |   |   |   |   |   |   |   
| figure |   |   |   |   |   |   |   |   |   |   |   
| math |   |   |   |   |   |   |   |   |   |   | 
| table |   |   |   |   |   |   |   |   |   |   |
| link      |   |   |   |   |   |   |   |   |   |   | 
| option |   |   |   |   |   |   |   |   |   |   | 
| progressbar |   |   |   |   |   |   |   |   |   |   |
| table |   |   |   |   |   |   |   |   |   |   | 
| slider |   |   |   |   |   |   |   |   |   |   |
| tab |   |   |   |   |   |   |   |   |   |   | 
| tabpanel |   |   |   |   |   |   |   |   |   |   | 
| tablist |   |   |   |   |   |   |   |   |   |   |
| treeitem |   |   |   |   |   |   |   |   |   |   |
| tree |   |   |   |   |   |   |   |   |   |   |
| treegrid |   |   |   |   |   |   |   |   |   |   |
| tooltip |   |   |   |   |   |   |   |   |   |   |

### Screen Reader / Browser Matrix

|            | NVDA      | JAWS      | VoiceOver | VoiceOver (iOS) | Narrator |
| :-------- | :--------: | :--------: | :--------: | :--------: | :--------: |
| Chrome (Windows)    |  ✔️   | ✔️    |          |          |          |
| Firefox (Windows)      |  ✔️   |          |          |          |          |
| Safari (Mac)                |          |          |  ✔️   |          |          |
| Chrome (Android)      |          |          |          |          |          |
| Chrome (iOS)             |          |          |          |          |          |
| Safari (iOS)                 |          |          |          |          |          |

#### Versions
Chrome 97.0.4692.99, Firefox 96.0.3, Safari 15.1 (17612.2.9.1.20)
NVDA 2021.3.1, JAWS 2022.2112.24 ILM Beginner, VoiceOver (all using default settings)
Windows 11 Build 22000.438, MacBook Pro/ macOS Monterey 12.0.1

### Branches
The philosophy with these branches is that they start very simple (with role=region | group | img) and then ramp up in complexity from there. Every branch uses the same core aria strategy: aria-labelledby combined with role. We only test whether aria-labelledby works with text elements as their source, not title or desc elements. We are not checking whether SVG can use HTML elements as their source either. We are keeping SVG stuff fully contained and HTML stuff fully contained. We are also not checking other role strategies like aria-roledescription.

|                           | svg | g | circle | path | rect | text |
| :-------------- | :-: | :-: | :-: | :-: | :-: | :-: |
| role                    | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ |
| aria-labelledby | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ |

#### "Region" Branch #25 
The simplest roles and layout. We are testing grouped elements inside an SVG and whether they have parity with a similar HTML structure.

|                          | svg | g | circle | path | rect | text |
| :---------------- | :-: | :-: | :-: | :-: | :-: | :-: |
| region              | ✔️ |       |       |       |       |       |
| group               |       | ✔️ |       |       |       |       |
| img                  |       |       | ✔️ | ✔️ | ✔️ | ✔️ |

#### "Graphics Document" Branch #24 
The next-simplest roles and layout. We are testing grouped elements inside an SVG and whether they have parity with a similar HTML structure. We are also testing if the fallback roles are helpful at all (or necessary).

|                                              | svg | g | circle | path | rect | text |
| :------------------------------ | :-: | :-: | :-: | :-: | :-: | :-: |
| graphics-document region   | ✔️ |       |       |       |       |       |
| graphics-object group          |       | ✔️ |       |       |       |       |
| graphics-symbol img            |       |       | ✔️ | ✔️ | ✔️ | ✔️ |

#### "Button" Branch #23  
We are now testing the simplest structure except instead of an img role on the children, they are interactive. We are using tabindex in this example for a simple keyboard-only navigation coverage test. We are using javascript to handle the state change on aria-pressed and aria-live elements.

|                             | svg | g | circle | path | rect | text |
| :------------------ | :-: | :-: | :-: | :-: | :-: | :-: |
| region                  | ✔️ |       |       |       |       |       |
| group                   |       | ✔️ |       |       |       |       |
| button                  |       |       | ✔️ | ✔️ | ✔️ | ✔️ |
| aria-pressed         |       |       | ✔️ | ✔️ | ✔️ | ✔️ |
| aria-live                |       |       |       |       |       | ✔️ |

#### "Menu" Branch #26   
This is the most complex branch. We are testing keyboard navigation with arrow keys (these are not intended to be perfectly designed for material honesty but rather just whether javascript and key interception works on SVG with screen readers). We are testing whether g elements can be interactive (as menuitems that open new sub menus) and then whether the children elements can be interactive as well. Due to the complexity of this example (cooking up keyboard navigation from scratch), we will want to be careful to keep this from technical scope creep. This branch is not intended to be an example of perfect javascript arrow-key interactivity but just whether javascript arrow-key interactivity can work at all.

|                             | svg | g | circle | path | rect | text |
| :------------------ | :-: | :-: | :-: | :-: | :-: | :-: |
| menu                        | ✔️ | ✔️ |       |       |       |       |
| menuitem                 |       | ✔️ |       |       |       |       |
| menuitemcheckbox  |       |       | ✔️ | ✔️ | ✔️ | ✔️ |
| aria-haspopup          |       | ✔️ |       |       |       |       |
| aria-expanded          |       | ✔️ |       |       |       |       |
| aria-checked             |       |       | ✔️ | ✔️ | ✔️ | ✔️ |

**Comment by shelbyelavsky** [comment]
### Current Testing Matrix Details
_The versions of all tools are noted in each issue's "Technical Details" section to account for updates during testing._ 

**Computers:**
Windows 11 Build 22000.438
MacBook Pro/ macOS Monterey 12.0.1

**Screen readers:**
JAWS 
NVDA
VoiceOver

**Browsers:**
Chrome Version
Firefox
Safari
